File: v4/src/ui.lua
local cards = require("src.cards")

local M = {}

local CARD_W = 70
local CARD_H = 100
local GAP = 10
local STACK_OFFSET = 18
local STARTER_X = 30
local STARTER_Y = 490

function M.set_font()
  local font = love.graphics.newFont(16)
  love.graphics.setFont(font)
end

function M.draw_card(card, x, y, highlighted)
  if highlighted then
    love.graphics.setColor(0.3, 0.6, 0.9, 0.4)
    love.graphics.rectangle("fill", x, y, CARD_W, CARD_H, 6, 6)
  end
  if card.vined then
    love.graphics.setColor(0.5, 0.6, 0.5)
  else
    love.graphics.setColor(1, 1, 1)
  end
  love.graphics.rectangle("line", x, y, CARD_W, CARD_H, 6, 6)
  if card.hidden then
    love.graphics.print("??", x + 10, y + 10)
  else
    love.graphics.print(cards.card_label(card), x + 10, y + 10)
  end
end

function M.draw_face_up_card(card, x, y)
  M.draw_card(card, x, y, false)
end

function M.draw_hand(hand, x, y, selected)
  for i = 1, #hand do
    local card = hand[i]
    local card_x = x + (i - 1) * (CARD_W + GAP)
    local is_selected = selected and selected[i]
    M.draw_card(card, card_x, y, is_selected)
    love.graphics.setColor(1, 1, 1)
    love.graphics.print(tostring(i), card_x + 30, y + CARD_H + 6)
  end
end

function M.draw_cards_row(cards_list, x, y)
  for i = 1, #cards_list do
    local card_x = x + (i - 1) * (CARD_W + GAP)
    M.draw_card(cards_list[i], card_x, y, false)
  end
end

function M.draw_combo_entries(entries, x, y)
  local cursor = y
  local row_height = CARD_H + 10
  for i = 1, #entries do
    local entry = entries[i]
    love.graphics.print(entry.label, x, cursor)
    if entry.cards and #entry.cards > 0 then
      M.draw_cards_row(entry.cards, x + 140, cursor - 10)
      cursor = cursor + row_height
    else
      cursor = cursor + 18
    end
  end
  return cursor
end

function M.draw_text_block(lines, x, y)
  for i = 1, #lines do
    love.graphics.print(lines[i], x, y + (i - 1) * 18)
  end
end

function M.draw_scoreboard(state)
  local lines = {
    "Level: " .. tostring(state.level),
    "Goal: 121",
    "Dealer: " .. state.dealer,
    "Player: " .. tostring(state.player_score),
    "AI: " .. tostring(state.ai_score),
  }
  M.draw_text_block(lines, 30, 20)
end

function M.draw_phase(phase, x, y)
  love.graphics.print("Phase: " .. phase, x, y)
end

function M.draw_peg_stack(stack, x, y)
  for i = 1, #stack do
    local card_x = x + (i - 1) * 40
    M.draw_card(stack[i], card_x, y, false)
  end
end

local function draw_face_down_card(x, y, alpha)
  love.graphics.setColor(0.12, 0.12, 0.18, alpha)
  love.graphics.rectangle("fill", x, y, CARD_W, CARD_H, 6, 6)
  love.graphics.setColor(1, 1, 1, alpha)
  love.graphics.rectangle("line", x, y, CARD_W, CARD_H, 6, 6)
end

local function card_alpha_for_index(cut, index)
  if cut.selected and index == cut.selected then
    if cut.status == "reveal" or cut.status == "clear_right" or cut.status == "slide" then
      return 0
    end
  end
  if cut.status == "aim" then
    return 1
  end
  if not cut.selected then
    return 1
  end
  if index < cut.selected then
    if cut.status == "clear_left" then
      return cut.fade_left or 0
    end
    return 0
  end
  if index > cut.selected then
    if cut.status == "clear_right" then
      return cut.fade_right or 0
    end
    if cut.status == "drop" then
      return 1
    end
    if cut.status == "clear_left" then
      return 1
    end
    return 0
  end
  return 1
end

function M.draw_cut_stack(cut, x, y)
  for i = 1, cut.visual_size do
    local alpha = card_alpha_for_index(cut, i)
    if alpha > 0 then
      local x_offset = (i - 1) * STACK_OFFSET
      local y_offset = 0
      if cut.selected and i == cut.selected and cut.status == "drop" then
        y_offset = y_offset + (cut.drop_offset or 0)
      end
      draw_face_down_card(x + x_offset, y + y_offset, alpha)
    end
  end

  if cut.selected and (cut.status == "reveal" or cut.status == "clear_right") then
    local base_x = x + (cut.selected - 1) * STACK_OFFSET
    local base_y = y + (cut.drop_offset or 0)
    M.draw_face_up_card(cut.card, base_x, base_y)
  elseif cut.selected and cut.status == "slide" then
    local base_x = x + (cut.selected - 1) * STACK_OFFSET
    local base_y = y + (cut.drop_offset or 0)
    local t = cut.slide_progress or 0
    local target_x = cut.target_x or STARTER_X
    local target_y = cut.target_y or STARTER_Y
    local draw_x = base_x + (target_x - base_x) * t
    local draw_y = base_y + (target_y - base_y) * t
    M.draw_face_up_card(cut.card, draw_x, draw_y)
  end

  if cut.status == "aim" then
    local pointer_x = x + (cut.pointer - 1) * STACK_OFFSET + CARD_W / 2
    love.graphics.setColor(1, 0.8, 0.2, 0.9)
    love.graphics.polygon("fill", pointer_x - 8, y - 12, pointer_x + 8, y - 12, pointer_x, y)
    love.graphics.setColor(1, 1, 1, 1)
  end
end

function M.card_dimensions()
  return CARD_W, CARD_H, GAP
end

function M.starter_position()
  return STARTER_X, STARTER_Y
end

return M

File: src/ui.ts
import { cardLabel } from "./cards.js";

export type Card = {
  rank: number;
  suit: number;
  hidden?: boolean;
  vined?: boolean;
  enhancement?: string;
};

const CARD_W = 70;
const CARD_H = 100;
const GAP = 10;
const STACK_OFFSET = 18;
const STARTER_X = 30;
const STARTER_Y = 490;

const CANVAS_W = 1200;
const CANVAS_H = 800;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function setColor(
  ctx: CanvasRenderingContext2D,
  r: number,
  g: number,
  b: number,
  a: number = 1
): void {
  const css = `rgba(${r * 255 | 0}, ${g * 255 | 0}, ${b * 255 | 0}, ${a})`;
  ctx.fillStyle = css;
  ctx.strokeStyle = css;
}

function roundRect(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  w: number,
  h: number,
  r: number
): void {
  const radius = Math.min(r, w / 2, h / 2);
  ctx.beginPath();
  ctx.moveTo(x + radius, y);
  ctx.lineTo(x + w - radius, y);
  ctx.arcTo(x + w, y, x + w, y + radius, radius);
  ctx.lineTo(x + w, y + h - radius);
  ctx.arcTo(x + w, y + h, x + w - radius, y + h, radius);
  ctx.lineTo(x + radius, y + h);
  ctx.arcTo(x, y + h, x, y + h - radius, radius);
  ctx.lineTo(x, y + radius);
  ctx.arcTo(x, y, x + radius, y, radius);
  ctx.closePath();
}

function printAligned(
  ctx: CanvasRenderingContext2D,
  text: string,
  x: number,
  y: number,
  w: number,
  align: "left" | "center" | "right"
): void {
  const measured = ctx.measureText(text).width;
  let drawX = x;
  if (align === "center") {
    drawX = x + (w - measured) / 2;
  } else if (align === "right") {
    drawX = x + w - measured;
  }
  ctx.fillText(text, drawX, y);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function setFont(ctx: CanvasRenderingContext2D): void {
  ctx.font = "16px monospace";
}

export function drawCard(
  ctx: CanvasRenderingContext2D,
  card: Card,
  x: number,
  y: number,
  highlighted: boolean
): void {
  if (highlighted) {
    setColor(ctx, 0.3, 0.6, 0.9, 0.4);
    roundRect(ctx, x, y, CARD_W, CARD_H, 6);
    ctx.fill();
  }

  if (card.vined) {
    setColor(ctx, 0.5, 0.6, 0.5);
  } else {
    setColor(ctx, 1, 1, 1);
  }
  roundRect(ctx, x, y, CARD_W, CARD_H, 6);
  ctx.stroke();

  if (card.hidden) {
    ctx.fillText("??", x + 10, y + 10 + 14); // offset by ~font ascent
  } else {
    ctx.fillText(cardLabel(card), x + 10, y + 10 + 14);
  }
}

export function drawFaceUpCard(
  ctx: CanvasRenderingContext2D,
  card: Card,
  x: number,
  y: number
): void {
  drawCard(ctx, card, x, y, false);
}

export function drawHand(
  ctx: CanvasRenderingContext2D,
  hand: Card[],
  x: number,
  y: number,
  selected: boolean[] | null
): void {
  for (let i = 0; i < hand.length; i++) {
    const card = hand[i];
    const cardX = x + i * (CARD_W + GAP);
    const isSelected = selected ? selected[i] : false;
    drawCard(ctx, card, cardX, y, isSelected);
    setColor(ctx, 1, 1, 1);
    ctx.fillText(String(i + 1), cardX + 30, y + CARD_H + 6 + 14);
  }
}

export function drawCardsRow(
  ctx: CanvasRenderingContext2D,
  cards: Card[],
  x: number,
  y: number
): void {
  for (let i = 0; i < cards.length; i++) {
    const cardX = x + i * (CARD_W + GAP);
    drawCard(ctx, cards[i], cardX, y, false);
  }
}

export type ComboEntry = {
  label: string;
  cards?: Card[];
};

export function drawComboEntries(
  ctx: CanvasRenderingContext2D,
  entries: ComboEntry[],
  x: number,
  y: number
): number {
  let cursor = y;
  const rowHeight = CARD_H + 10;
  for (const entry of entries) {
    ctx.fillText(entry.label, x, cursor + 14);
    if (entry.cards && entry.cards.length > 0) {
      drawCardsRow(ctx, entry.cards, x + 140, cursor - 10);
      cursor += rowHeight;
    } else {
      cursor += 18;
    }
  }
  return cursor;
}

export function drawTextBlock(
  ctx: CanvasRenderingContext2D,
  lines: string[],
  x: number,
  y: number
): void {
  for (let i = 0; i < lines.length; i++) {
    ctx.fillText(lines[i], x, y + i * 18 + 14);
  }
}

export type ScoreboardState = {
  level: number | string;
  dealer: string;
  player_score: number;
  ai_score: number;
};

export function drawScoreboard(
  ctx: CanvasRenderingContext2D,
  state: ScoreboardState
): void {
  const lines = [
    "Level: " + String(state.level),
    "Goal: 121",
    "Dealer: " + state.dealer,
    "Player: " + String(state.player_score),
    "AI: " + String(state.ai_score),
  ];
  drawTextBlock(ctx, lines, 30, 20);
}

export function drawPhase(
  ctx: CanvasRenderingContext2D,
  phase: string,
  x: number,
  y: number
): void {
  ctx.fillText("Phase: " + phase, x, y + 14);
}

export function drawPegStack(
  ctx: CanvasRenderingContext2D,
  stack: Card[],
  x: number,
  y: number
): void {
  for (let i = 0; i < stack.length; i++) {
    const cardX = x + i * 40;
    drawCard(ctx, stack[i], cardX, y, false);
  }
}

// ---------------------------------------------------------------------------
// Internal helper – face-down card
// ---------------------------------------------------------------------------

function drawFaceDownCard(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  alpha: number
): void {
  setColor(ctx, 0.12, 0.12, 0.18, alpha);
  roundRect(ctx, x, y, CARD_W, CARD_H, 6);
  ctx.fill();
  setColor(ctx, 1, 1, 1, alpha);
  roundRect(ctx, x, y, CARD_W, CARD_H, 6);
  ctx.stroke();
}

export type CutStack = {
  visual_size: number;
  selected?: number;
  status: string;
  fade_left?: number;
  fade_right?: number;
  drop_offset?: number;
  slide_progress?: number;
  target_x?: number;
  target_y?: number;
  pointer: number;
  card: Card;
};

function cardAlphaForIndex(cut: CutStack, index: number): number {
  if (cut.selected !== undefined && index === cut.selected) {
    if (
      cut.status === "reveal" ||
      cut.status === "clear_right" ||
      cut.status === "slide"
    ) {
      return 0;
    }
  }
  if (cut.status === "aim") {
    return 1;
  }
  if (cut.selected === undefined) {
    return 1;
  }
  if (index < cut.selected) {
    if (cut.status === "clear_left") {
      return cut.fade_left ?? 0;
    }
    return 0;
  }
  if (index > cut.selected) {
    if (cut.status === "clear_right") {
      return cut.fade_right ?? 0;
    }
    if (cut.status === "drop") {
      return 1;
    }
    if (cut.status === "clear_left") {
      return 1;
    }
    return 0;
  }
  return 1;
}

export function drawCutStack(
  ctx: CanvasRenderingContext2D,
  cut: CutStack,
  x: number,
  y: number
): void {
  for (let i = 1; i <= cut.visual_size; i++) {
    const alpha = cardAlphaForIndex(cut, i);
    if (alpha > 0) {
      const xOffset = (i - 1) * STACK_OFFSET;
      let yOffset = 0;
      if (
        cut.selected !== undefined &&
        i === cut.selected &&
        cut.status === "drop"
      ) {
        yOffset += cut.drop_offset ?? 0;
      }
      drawFaceDownCard(ctx, x + xOffset, y + yOffset, alpha);
    }
  }

  if (
    cut.selected !== undefined &&
    (cut.status === "reveal" || cut.status === "clear_right")
  ) {
    const baseX = x + (cut.selected - 1) * STACK_OFFSET;
    const baseY = y + (cut.drop_offset ?? 0);
    drawFaceUpCard(ctx, cut.card, baseX, baseY);
  } else if (cut.selected !== undefined && cut.status === "slide") {
    const baseX = x + (cut.selected - 1) * STACK_OFFSET;
    const baseY = y + (cut.drop_offset ?? 0);
    const t = cut.slide_progress ?? 0;
    const targetX = cut.target_x ?? STARTER_X;
    const targetY = cut.target_y ?? STARTER_Y;
    const drawX = baseX + (targetX - baseX) * t;
    const drawY = baseY + (targetY - baseY) * t;
    drawFaceUpCard(ctx, cut.card, drawX, drawY);
  }

  if (cut.status === "aim") {
    const pointerX = x + (cut.pointer - 1) * STACK_OFFSET + CARD_W / 2;
    setColor(ctx, 1, 0.8, 0.2, 0.9);
    ctx.beginPath();
    ctx.moveTo(pointerX - 8, y - 12);
    ctx.lineTo(pointerX + 8, y - 12);
    ctx.lineTo(pointerX, y);
    ctx.closePath();
    ctx.fill();
    setColor(ctx, 1, 1, 1, 1);
  }
}

export function cardDimensions(): { w: number; h: number; gap: number } {
  return { w: CARD_W, h: CARD_H, gap: GAP };
}

export function starterPosition(): { x: number; y: number } {
  return { x: STARTER_X, y: STARTER_Y };
}
